arangod/Replication2/ReplicatedLog/ILogParticipant.h

```diff
@@ -0,0 +1,84 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2021-2021 ArangoDB GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Tobias Gödderz
+////////////////////////////////////////////////////////////////////////////////
+
+#pragma once
+
+#include "Replication2/DeferredExecution.h"
+#include "Replication2/ReplicatedLog/LogCommon.h"
+#include "Replication2/ReplicatedLog/ReplicatedLogMetrics.h"
+#include "Replication2/ReplicatedLog/types.h"
+
+#include <Futures/Future.h>
+#include <Futures/Promise.h>
+
+#include <map>
+#include <memory>
+
+namespace arangodb::replication2::replicated_log {
+
+struct LogCore;
+struct LogStatus;
+
+/**
+* @brief Interface for a log participant: That is, usually either a leader or a
+* follower (LogLeader and LogFollower). Can also be a LogUnconfiguredParticipant,
+* e.g. during startup.
+* The most prominent thing this interface provides is that each instance is
+* responsible for a singular LogCore, which can be moved out with resign().
+*/
+struct ILogParticipant {
+  [[nodiscard]] virtual auto getStatus() const -> LogStatus = 0;
+  virtual ~ILogParticipant() = default;
+  [[nodiscard]] virtual auto resign() &&
```

Comment:
While `resign()` should only be called on an rvalue, there needs to be some sort of external synchronization to make it thread-safe.
Otherwise it is possible that some thread is currently executing inside one of the LogParticipant's method, while another thread is calling `resign()`. This wouldn't be thread-safe per se.
I guess there is some other piece of code (haven't reviewed fully) that ensures no other thread can be busy in a LogParticipant while `resign()` is called. If so, it may be useful to add a description to the interface.

Reply:
`resign()` is explicitly designed to *not* need external synchronization, in the following sense:

The participant makes sure all access to its LogCore is concurrency-safe, particularly in relation to `resign()`. If the LogCore is accessed after the participant resigned, an exception with error code `TRI_ERROR_REPLICATION_REPLICATED_LOG_LEADER_RESIGNED` or `TRI_ERROR_REPLICATION_REPLICATED_LOG_FOLLOWER_RESIGNED` is thrown.

If you find that explanation satisfactory, I'll add a comment with it.